<span style='color:brown'>Python es un lenguaje de programación ampliamente útil y orientado a objetos. Su sentido significativo junto a sus increíbles bibliotecas le han dado el respeto de ser el idioma preferido para carreras como la de un científico de datos o un experto en IA.</span>

<span style='color:brown'>Python también es elegido con frecuencia como el lenguaje para la programación en escuelas y universidades debido a su facilidad de aprendizaje. Estos tutoriales instructivos sirven para ayudar a aprender Python desde cero y hasta un nivel avanzado en este lenguaje.</span>

## Lección 1 - Introducción a la programación con Python

#### Primeros pasos con Python y Jupyter

![Operadores](Basic-Operators-in-Python-blue-2.jpg)

### Parte 1 de "Análisis de datos con Python

Esta serie de tutoriales es una introducción amigable para principiantes a la programación y el análisis de datos utilizando el lenguaje de programación python. Estos tutoriales toman un enfoque práctico y centrado en la codificación. La mejor manera de aprender el material es ejecutar el código y experimentar con él usted mismo. Echa un vistazo a la serie completa aquí:

>- Primeros pasos con Python y Jupyter
>- Un recorrido rápido por variables y tipos de datos
>- Bifurcación mediante instrucciones y bucles condicionales
>- Escribir código reutilizable mediante funciones
>- Leer y escribir a archivos
>- Computación numérica con Python y Numpy
>- Análisis de datos tabulares con pandas
>- Visualización de datos mediante Matplotlib & Seaborn
>- Análisis exploratorio de datos - Un caso de estudio

En este tutorial se tratan los siguientes temas:

>- Realización de operaciones aritméticas con Python
>- Solución de problemas de varios pasos mediante variables
>- Evaluación de condiciones mediante Python
>- Combinación de condiciones con operadores lógicos
>- Adición de estilos de texto mediante Markdown

### Realización de operaciones aritméticas con Python
Empecemos por usar Python como una calculadora. Puedes escribir y ejecutar Python mediante una celda del tipo `código` dentro de tu cuaderno de Jupyter.

<ins>Trabajar con celdas</ins>: Para crear una nueva celda dentro de Jupyter, puedes seleccionar "Insertar > Insertar celda a continuación" en la barra de menús o simplemente presione el botón "+" en la barra de herramientas.
También puedes usar el método abreviado de teclado para crear una nueva celda (presionando la letra "a" para insertar una celda anterior a la seleccionada o "b" para insertar una celda después). Una vez creada una celda, haga clic en la misma para seleccionarla, a continuación, puede cambiar el tipo de celda a código o markdown (texto) mediante la opción de menú "Celda > Tipo de celda".
Para editar el contenido la celda haga doble clic en la misma.
Para aplicar los cambios y ejecutar una celda, utilice la opción de menú "Celda > Ejecutar celdas" o haga clic en el botón "Ejecutar" de la barra de herramientas o simplemente utilice el método abreviado de teclado ("Ctrl + Intro" o "Shift + Intro").
Puedes ver una lista completa de métodos abreviados de teclado mediante la opción de menú "Ayuda > Métodos abreviados de teclado".

### Realización de operaciones aritméticas con Python
Comencemos por usar Python como una calculadora. Puedes escribir y ejecutar Python mediante una celda de código dentro de Jupyter.

Ejecute las celdas de código siguientes para realizar cálculos y ver su resultado. Intente cambiar los números y vuelva a ejecutar las celdas modificadas para ver los resultados actualizados. ¿Puedes adivinar para qué se utilizan los operadores `// % **`?

<span style="color:brown">_Por cierto, utilizamos el símbolo "#" para inserir comentarios en nuestras celdas de código_</span>

In [47]:
# Esto es un comentario, no será ejecutado en nuestro código

In [48]:
# Suma
2 + 3 + 9

14

In [49]:
# Resta
99 -73

26

In [50]:
# Multiplicación
23.54 * -1432

-33709.28

In [51]:
# División
100 / 7

14.285714285714286

In [52]:
# Parte entera de la División
100 // 7

14

In [53]:
# Resto de la división
100 % 7

2

In [54]:
# Potenciación
5 ** 3

125

Como era de esperar, los operadores tienen prioridad sobre otros operadores como y según convenciones matemáticas. Puedes utilizar paréntesis para separar y ordenar cómo se realizan las operaciones. `( ) ** * / % // + -`

In [55]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

#### Solución de problemas de varios pasos mediante variables

Intentemos resolver el siguiente problema:

Una tienda de comestibles vende una bolsa de hielo por $1.25 y obtiene un beneficio del 20%. Si vende 500 bolsas de hielo, ¿cuánto beneficio total obtienen?

Podemos enumerar la información proporcionada y convertir gradualmente la palabra problema en una expresión matemática que se puede evaluar mediante Python.

Coste de la bolsa de hielo ($) = 1.25

Margen de beneficio = 20% = .2

Beneficio por bolsa ($) = margen de beneficio * costo de la bolsa de hielo = .2 * 1.25

Número de bolsas = 500

Beneficio total = número de bolsas * beneficio por bolsa = 500 * (.2 * 1.25)

In [56]:
500 * (.2 * 1.25)

125.0

Por lo tanto, la tienda de comestibles obtiene un beneficio total de $125. Si bien esta es una manera razonable de resolver un problema, no está del todo claro mirando la celda de código lo que representan los números. Podemos dar nombres a cada uno de los números mediante la creación de variablesde Python.

Mientras se trabaja con un lenguaje de programación como Python, la información se almacena en **variables**. Puede pensar en variables como contenedores para almacenar datos. Los datos almacenados dentro de una variable se denominan su valor.

In [57]:
# Declaremos las variables
coste_bolsa_hielo = 1.25
margen_beneficio = .2
num_bolsas = 500

Las variables ahora contienen los valores proporcionados. Podemos comprobar el valor de una variable escribiendo su nombre en una celda. También podemos combinar variables mediante operaciones aritméticas para crear otras variables.

Pista: al escribir el nombre de una variable existente en una celda de código dentro de Jupyter, simplemente escriba los primeros caracteres y presione la tecla "Tab" para autocompletar el nombre de la variable. Intente escribir una celda de código a continuación y pulse para autocompletarla.

In [58]:
coste_bolsa_hielo

1.25

In [59]:
margen_beneficio

0.2

In [60]:
num_bolsas

500

In [61]:
# Creamos/declaramos la variable margen_benef_por_bolsa
margen_benef_por_bolsa = margen_beneficio * coste_bolsa_hielo

In [62]:
margen_benef_por_bolsa

0.25

In [63]:
# Creamos/declaramos la variable margen_total
margen_total = num_bolsas * margen_benef_por_bolsa

In [64]:
margen_total

125.0

Si intentas ver el valor de una variable que no se ha definido antes, es decir, dado un valor mediante la instrucción, Python mostrará un error (name 'nombre_variable' is not defined)

In [65]:
Margen_neto

NameError: name 'Margen_neto' is not defined

Almacenar y manipular datos utilizando variables con el nombre adecuado es una excelente manera de explicar lo que hace el código.

Función `PRINT`: Para exibir la información o resultado de una celda utilizamos la función `print`.

Funciones: Una función es un conjunto reutilizable de instrucciones. Toma una o más entradas, realiza ciertas operaciones y a menudo devuelve una salida. Python proporciona muchas funciones integradas, pero también nos permite definir nuestras propias funciones.

In [ ]:
print("La tienda de comestibles obtuvo un beneficio total de $", margen_total)

Función **PRINT**: La función `PRINT` se utiliza para mostrar información. Se necesitan una o más entradas, que pueden ser texto (entre comillas simples `'` o duplas `"`), números, variables, expresiones matemáticas, etc. Veremos más adelante sobre variables y funciones en el siguiente tutorial.

Crear una celda de código para cada operación de variable o operación matemática puede volverse tedioso. Afortunadamente, Jupyter nos permite escribir varias líneas de código dentro de una sola celda de código y ejecutarlos todos de una sola vez.

Reescribamos la solución a nuestro problema de palabras dentro de una sola celda.

In [66]:
# Declaremos las variables
coste_bolsa_hielo = 1.25
margen_beneficio = .2
num_bolsas = 500

# Declaremos las variables de resultado realizando los cálculos con las variables anteriores
margen_benef_por_bolsa = margen_beneficio * coste_bolsa_hielo
margen_total = num_bolsas * margen_benef_por_bolsa

# Mostremos en pantalla el resultado
print("La tienda de comestibles obtuvo un beneficio total de $", margen_total)

La tienda de comestibles obtuvo un beneficio total de $ 125.0


Recuerda que estamos usando el carácter # para agregar comentarios dentro de nuestro código.

**Comentarios**: Los comentarios y las líneas en blanco se ignoran durante la ejecución, pero son útiles para proporcionar información a los seres humanos (incluido usted mismo) sobre lo que hace el código. Los comentarios pueden estar en línea (al final de algún código), en una línea independiente o incluso abarcar varias líneas.

Los comentarios en línea y de una sola línea comienzan con #, mientras que los comentarios multilínea comienzan y terminan con tres comillas, es decir. Ejemplo: '''Este es un cometario de varias lineas'''.

In [ ]:
mi_num_pref = 9 # Este es un comentario de ua sola línea

In [ ]:
# Este también es un comentario de una sola línea
mi_peor_num = 1

"""Este es un comentario de varias líneas.
Escribe lo poco o lo mucho que quieras.

Los comentarios son realmente útiles para la gente que lee tu código, pero trate de mantenerlos cortos y al punto.

Además, si usas buenos nombres de variables, tu código suele ser un código que se explica por sí mismo, ¡y puede que ni siquiera necesites comentarios!
"""
a_número_neutral = 5

#### Evaluación de condiciones mediante Python
Además de las operaciones aritméticas, Python también proporciona varias operaciones para comparar números y variables.

![Operadores_comparacion](Operadores_comparacion.png)

El resultado de una operación de comparación será siempre `True` o `False` (tenga en cuenta las mayúsculas T y F). Estas son palabras clave especiales en Python. Probemos algunos experimentos con operadores de comparación.

In [ ]:
mi_num_fav = 1
mi_ultim_num_fav = 5
mi_num_neutral = 3

In [ ]:
# Operación Igual que - True
mi_num_fav == 1

In [ ]:
# Operación Igual que - False
mi_num_fav == mi_ultim_num_fav

In [ ]:
# Operación de Distinto que - True
mi_num_fav != mi_num_neutral

In [ ]:
# Operación de Distinto que - False
mi_num_neutral != 3

In [ ]:
# Operación de Mayor que - True
mi_ultim_num_fav > mi_num_neutral

In [ ]:
# Operación de Mayor que - False
mi_num_fav > mi_ultim_num_fav

In [ ]:
# Operación Menor que - True
mi_num_fav < 10

In [ ]:
# Operación Menor que - False
mi_ultim_num_fav < mi_num_fav

In [ ]:
# Operación Mayor o igual que - True
mi_num_fav >= 1

In [ ]:
# Operación Mayor o igual que - False
mi_num_fav >= 3

In [ ]:
# Operación Menor o igual que - True
3 + 6 <= 9

In [ ]:
# Operación Menor o igual que - False
mi_num_fav + mi_num_neutral <= 3

Al igual que las operaciones aritméticas, el resultado de una operación de comparación también se puede almacenar en una variable.

In [ ]:
coste_bolsa_hielo = 1.25
es_la_bolsa_de_hielo_cara = coste_bolsa_hielo >= 10
print("Está la bolsa de hielo cara?", es_la_bolsa_de_hielo_cara)

#### Combinación de condiciones con operadores lógicos
Los operadores lógicos `and`, `or` y `not` operan según condiciones y valores `True` y `False` (también conocidos como booleanos). `and` y `or` operan en dos condiciones, mientras que `not` opera en una sola condición.

El operador retornará `True` cuando ambas condiciones se evalúan como `True`. De lo contrario, retornará `False`.

![Operadores_booleanos](Operadores_booleanos.jpg)

In [ ]:
mi_num_neutral = 3

In [ ]:
mi_num_neutral == 3 or mi_num_fav > 0

In [ ]:
mi_num_neutral != 3 or mi_num_fav > 0

In [ ]:
mi_num_fav < 0 or True

In [ ]:
False or False

El operador `not` devuelve `False` si una condición es `True` y `True` si la condición es `False`.

In [ ]:
not mi_num_neutral == 3

In [ ]:
not mi_num_fav < 0

In [ ]:
not False

In [ ]:
not True

Logical operators can be combined to form complex conditions. Use round brackets or parentheses ( and ) to indicate the order in which logical operators should be applied.
Los operadores lógicos se pueden combinar para formar condiciones complejas. Utilice corchetes o paréntesis para indicar el orden en que se deben aplicar los operadores lógicos.

In [ ]:
(2 > 3 and 4 <= 5) or not (mi_num_fav < 0 and True)

In [ ]:
not (True and 0 < 1) or (False and True)

Si no se utilizan paréntesis, los operadores lógicos se aplican de izquierda a derecha.

In [ ]:
not True and 0 < 1 or False and True

Experimente con operadores aritméticos, condicionales y lógicos en Python utilizando la naturaleza interactiva de los cuadernos jupyter. Aprenderemos más sobre variables y funciones más adelante.

#### Adición de estilos de texto mediante Markdown
Agregar explicaciones usando celdas de texto (como esta) es una gran manera de hacer que tu cuaderno sea informativo para otros lectores. También es útil si necesitas volver a él en el futuro. Puede hacer doble clic en una celda de texto dentro de Jupyter para editarla. En el modo de edición, observará que el texto se ve ligeramente diferente (por ejemplo, el encabezado tiene un `##` como prefijo). Este texto se escribe con Markdown, una forma sencilla de agregar estilos al texto. Ejecute esta celda para ver la salida sin los caracteres especiales. Puede cambiar de un lado a otro entre la fuente y la salida para aplicar un estilo específico.

Por ejemplo, puedes utilizar uno o varios `#` al inicio de una línea para crear encabezados de diferentes tamaños:

# Titulo
## Titulo
### Titulo
#### Titulo
##### Titulo
###### Titulo

Para crear una lista con viñetas o numerada, simplemente inicie una línea con `*` o `1`.

Una lista con viñetas:
* Punto 1
* Punto 2
* Punto 3

Una lista numerada:
1.manzana
2.Plátano
3.Piña

Puedes insertar texto en negrita utilizando doble asterisco, por ejemplo, **algún texto en negrita**, o hacerlo cursivo utilizando un solo asterisco, por ejemplo, *algún texto cursivo*. También puede crear enlaces, insertar imágenes, etc..